pkg/engine/deployment.go

```diff
@@ -181,12 +183,24 @@ func newDeployment(ctx *Context, info *deploymentContext, opts deploymentOptions
 	if err != nil {
 		return nil, err
 	}
-	plugctx = plugctx.WithCancelChannel(ctx.Cancel.Canceled())
+
+	// Keep the plugin context open until the context is terminated, to allow for graceful provider cancellation.
+	plugctx = plugctx.WithCancelChannel(ctx.Cancel.Terminated())
+
+	// Set up a goroutine that will signal cancellation to the source if the caller context
+	// is cancelled.
+	cancelCtx, cancelFunc := context.WithCancel(context.Background())
+	go func() {
+		<-ctx.Cancel.Canceled()
```

Comment:
Are we sure that `ctx.Cancel.Canceled` will be called eventually? We don't want to leak the go routine.
CC @Frassle

Reply:
It does seem like it is closed eventually, see [here](https://github.com/pulumi/pulumi/blob/f23b83df97d1dc1975fe8867baa1083eeddbbcd6/pkg/backend/httpstate/backend.go#L1284). I believe the call to `plugctx.WithCancelChannel` would likewise leak a goroutine.

Reply:
Why do we even need a new context here? `ctx.Cancel.Context()` should cover it, that will trigger when canceled or terminated.

Reply:
You're right, my rationale was simply to emit the log line. Do tell if you feel strongly.

Reply:
Not strongly, log line could be helpful.

Reply:
I tried to make this change, but `ctx.Cancel` doesn't seem to directly provide a `context.Context` ([ref](https://github.com/pulumi/pulumi/blob/f23b83df97d1dc1975fe8867baa1083eeddbbcd6/pkg/util/cancel/context.go#L85)).